src/box/sql/show.c

```diff
@@ -0,0 +1,415 @@
+{
+	char *buf = &desc->buf[desc->len];
+	size_t size = desc->size - desc->len;
+	va_list ap;
+	va_start(ap, fmt);
+	int rc = vsnprintf(buf, size, fmt, ap);
+	va_end(ap);
+	assert(rc >= 0);
+	if ((size_t)rc < size)
+		goto end;
+
+	while ((size_t)rc + desc->len >= desc->size)
+		desc->size *= 2;
+	desc->buf = sql_xrealloc(desc->buf, desc->size);
+	buf = &desc->buf[desc->len];
+	size = desc->size - desc->len;
+	va_start(ap, fmt);
+	rc = vsnprintf(buf, size, fmt, ap);
+	va_end(ap);
+end:
+	desc->len += rc;
+	assert(rc >= 0 && desc->len < desc->size);
+}
+
+/** Append a new error to the object used to accumulate a statement. */
+static void
+sql_err_append(struct sql_desc *desc, const char *type, const char *name)
+{
+	const char *fmt;
+	if (desc->err_len == 0)
+		fmt = "/* %s \"%s\" cannot be defined in this statement. */";
+	else
+		fmt = "\n/* %s \"%s\" cannot be defined in this statement. */";
+	char *buf = &desc->err_buf[desc->err_len];
+	size_t size = desc->err_size - desc->err_len;
+	int rc = snprintf(buf, size, fmt, type, name);
+	assert(rc >= 0);
+	if ((size_t)rc < size)
+		goto end;
+
+	while ((size_t)rc + desc->err_len >= desc->err_size)
+		desc->err_size *= 2;
+	desc->err_buf = sql_xrealloc(desc->err_buf, desc->err_size);
+	buf = &desc->err_buf[desc->err_len];
+	size = desc->err_size - desc->err_len;
+	rc = snprintf(buf, size, fmt, type, name);
+end:
+	desc->err_len += rc;
+	assert(rc >= 0 && desc->err_len < desc->err_size);
+}
+
+/** Finalize a described statement. */
+static char *
+sql_desc_finalize(struct sql_desc *desc)
+{
+	if (desc->err_len > 0) {
+		if (desc->len > 0)
+			sql_desc_append(desc, "\n");
+		sql_desc_append(desc, desc->err_buf);
+	}
+	char *buf = desc->buf;
+	sql_xfree(desc->err_buf);
+	memset(desc, 0, sizeof(*desc));
+	return buf;
+}
+
+/** Add a field foreign key constraint to the statement description. */
+static void
+sql_describe_field_foreign_key(struct sql_desc *desc,
+			       const struct tuple_constraint_def *cdef)
+{
+	assert(cdef->type == CONSTR_FKEY && cdef->fkey.field_mapping_size == 0);
+	const struct space *foreign_space = space_by_id(cdef->fkey.space_id);
+	assert(foreign_space != NULL);
+	const struct tuple_constraint_field_id *field = &cdef->fkey.field;
+	if (field->name_len > 0) {
+		uint32_t id;
+		uint32_t hash = field_name_hash(field->name, field->name_len);
+		if (tuple_fieldno_by_name(foreign_space->def->dict, field->name,
+					  field->name_len, hash, &id) != 0) {
+			sql_err_append(desc, "Foreign key", cdef->name);
+			return;
+		}
+	} else if (field->id >= foreign_space->def->field_count) {
+		sql_err_append(desc, "Foreign key", cdef->name);
+		return;
+	}
+	sql_desc_append(desc, "", cdef->name);
+	const char *field_name = field->name_len != 0 ? field->name :
+				 foreign_space->def->fields[field->id].name;
+	sql_desc_append(desc, " CONSTRAINT \"%s\" REFERENCES \"%s\"(\"%s\")",
+			cdef->name, foreign_space->def->name, field_name);
+}
+
```

Comment:
I would say this behaviour is quite unexpected:
```
tarantool> box.execute('SHOW CREATE TABLE "404"')
---
- metadata:
  - name: CREATE TABLE statement
    type: string
  rows: []
...
```
```
tarantool> s = box.schema.create_space('test')
---
...

tarantool> box.execute('SHOW CREATE TABLE "test"')
---
- metadata:
  - name: CREATE TABLE statement
    type: string
  rows:
  - ['/* Space "test" cannot be defined in this statement. */']
...
```
Why aren't these normal errors? Users have no sane way to see if the statement has failed. They would need to check in their code for no rows and for `/*` in the beginning of the statement?

Reply:
I don't think the statement failed in the second case. As for the first one, since we didn't find anything, it seems logical to me that we return empty result. Either way, let's ask @unera.

Reply:
But this is not a `"CREATE TABLE"` statement. So the request couldn't return what was expected. That is an error. You even literally call this message in your code `/** Buffer for errors. */`.

Reply:
Replaced error messages by actual errors. However, this actually affects `SHOW CREATE TABLE;` statement, since one wrong space and no information will be given. Maybe we allow error messages for all cases except `Space 'space_name' does not exist`?

Replaced error messages with real errors. However, this actually affects the `SHOW CREATE TABLE;` statement, as one wrong space and no proper result will be provided. Maybe we allow error messages for all cases except `Space 'space_name' does not exist`?

Reply:
I agree, unknown name should throw an error.

Reply:
update: we should generate an error only if `SHOW CREATE TABLE t` links to unknown (or inaccessible) `t`.

The operator can contain some comments:
- about indexes, that can not be created by SQL
- about any the other problems

In an extreme case, this statement will only contain comments